pkg/backends/promremotewriter/promremotewriter.go

```diff
@@ -0,0 +1,381 @@
+package promremotewriter
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"github.com/tilinna/clock"
+
+	"github.com/atlassian/gostatsd"
+	"github.com/atlassian/gostatsd/internal/util"
+	"github.com/atlassian/gostatsd/pb"
+	"github.com/atlassian/gostatsd/pkg/stats"
+	"github.com/atlassian/gostatsd/pkg/transport"
+)
+
+const (
+	// BackendName is the name of this backend.
+	BackendName                  = "promremotewriter"
+	defaultUserAgent             = "gostatsd" // TODO: Add version
+	defaultMaxRequestElapsedTime = 15 * time.Second
+	// defaultMetricsPerBatch is the default number of metrics to send in a single batch.
+	defaultMetricsPerBatch = 1000
+	// maxResponseSize is the maximum response size we are willing to read.
+	maxResponseSize = 1024
+)
+
+var (
+	// defaultMaxRequests is the number of parallel outgoing requests to prometheus.  As this mixes
+	// both CPU (TLS) and network bound operations, balancing may require some experimentation.
+	defaultMaxRequests = uint(10 * runtime.NumCPU())
+)
+
+// Client represents a Prometheus Remote Writer client.
+type Client struct {
```

Comment:
Is there an interface that this implements?

Reply:
Yup, `gostatsd.SendMetricsAsync`.  It's not formally checked, but the code won't compile if it doesn't implement it (`NewClientFromViper` will fail its return type)

Reply:
Not worth doing the compiler time interface check within the file?

Reply:
We do, `NewClientFromViper` will fail if it doesn't match.